### ISSUE TITLE
Use aliased imports in the gallery

### ISSUE THREAD

**Issue by jaccovanschaik** [has2k1/plotnine#24]
I would like to suggest using "import plotnine as p9" (or something like that) in the gallery code instead of "from plotnine import *", and then using "p9" as a prefix for all identifiers defined by plotnine. That would make it much clearer which of them come from plotnine and which don't.

**Comment by shoyer** [comment]
Another option worth considering is using namespaces as part of the API to replace the underscores from ggplot2. For example, instead of
```python
from plotnine import *

(ggplot(...) + 
 facet_wrap(...) +
 geom_tile(...))
```
you could write:
```
from plotnine import ggplot, facet, geom

(ggplot(...) + 
 facet.wrap(...) +
 geom.tile(...))
```

R isn't as strict about namespaces, so this sort of API makes less there.

**Comment by chris-b1** [comment]
I like @shoyer's idea, but another alternative I'll throw out is supporting more of a fluent, chain-able api.  It would break the direct translation from ggplot, but does feel more "pythonic" to me, for what that's worth.

```python
from plotnine import ggplot

(ggplot(...)
 .facet_wrap(...)
 .geom_tile(...))
```

**Comment by azag0** [comment]
@chris-b1 You can quite easily achieve this yourself:

```python
import plotnine

class _ggplot(ggplot):
    def __getattr__(self, attr):
        def f(*args, _self=self, **kwargs):
            return _self + getattr(plotnine, attr)(*args, **kwargs)
        return f

(_ggplot(...)
 .facet_wrap(...)
 .geom_tile(...))
```

**Comment by chris-b1** [comment]
Thanks @azag0  that's nice!  I added a `__dir__` so tab completion works too.
```python
def __dir__(self):
    return plotnine.__all__
```

**Comment by azag0** [comment]
Nice. Btw, going the opposite direction, from dot notation to plus notation, would be less straightforward, so I'd say that the current state is more general.

**Comment by has2k1** [comment]
Great. The solution is rather irrefutable. Thanks for the brainstorming.

**Comment by ChickenProp** [comment]
Ah, I was going to start a new issue to talk about this, but might as well keep the discussion here.

(We've [briefly spoken](https://news.ycombinator.com/item?id=14670096) on HN.)

I've spent a few months working on my own ggplot2-for-python, and then I discovered plotnine. But I do think I had a more natural API, pretty close to the one discussed here. I've mostly ported it as a shim on top of plotnine to demonstrate, [here are some examples](https://github.com/ChickenProp/gragrapy/blob/plotnine/pythonplot-examples.ipynb). Notably:

* Not expected to `import *`.
* More hierarchical namespace, `gg.geom.point` instead of `geom_point`.
* You can add to a plot with functions like `plot.geom('point')`, `plot.stat('smooth')`.
* These can be called in various ways: `plot.geom('point', color='red')`; `plot.geom(gg.geom.point, color='red')`; `plot.geom(gg.geom.point(color='red'))` all work.
* (For scales, it's `plot.scale('x', 'continuous', limits=(3,5))`)
* When you do use addition, lists and classes are accepted. So e.g. `plot + [gg.geom.point, gg.stat.smooth]`.
* Datasets included with the import (`gg.data.diamonds`), so no need to bring them in separately. Datasets are lazily loaded when used.
* (This one I didn't think I could shim on top of p9.) Instead of using dots like `color='..density..'` for calculated aesthetics, use `stat_color='density'`. I think this is more natural, and it has the advantage that you can set a mapping for both the initial data and the stat. E.g. `stat.smooth(aes(y=something, stat_y='ymax'))` will draw a smooth curve with the main line along the top edge of the ribbon. Admittedly, I don't think I've found a realistic use case for this.

I do like the `plot.geom_point()` suggestion in this thread as well. I'm not sure there's much to recommend one over the other.

(I will add as a sidenote that `gg` seems like an easier prefix to type than `p9`.)

**Comment by has2k1** [comment]
@ChickenProp, it is interesting see someone tryout customisations, in the notebook examples I had to double-check to make sure it was `plotnine` underneath.

Here are is current state of mind, roughly organised around the bullet points:

1. `import *` : This is the big issue and I am suprised not many have complained about it. It is very convenient, I got used to it and forgot about the demerits. I guess while working in Jupyter notebooks, one can have lower the standards of their code.

2. The two ways to solve the `import *` issues are either with hierarchical namespacing and/or method chaining. In trying to stay closer to the ggplot2 API, I prefer method chaining without any hierarchical namespacing. Adding to the ggplot object has to stay as part of the API. Here is how the options compare against each other:

   ```python
   # 0.
   # - Requires many imports
   # - No modifications
   (ggplot(...)
    + facet_wrap(...)
    + geom_tile(...))

   # 1.
   # - Requires many imports
   # - A lot of refactoring
   # - Potential namespace complications
   (ggplot(...)
    + facet.wrap(...)
    + geom.tile(...))

   # 2.
   # - Requires a single import
   # - Little refactoring
   (ggplot(...)
    .facet_wrap(...)
    .geom_tile(...))

   # 3.
   # - Requires a single import
   # - A lot of refactoring
   # - Potential namespace complications
   (ggplot(...)
    .facet.wrap(...)
    .geom.tile(...))

   ```

   The *potential namespace complications* points to issues with extending the package. As it stands, if you create a third-party `geom`, `stat`, ... as long as you inherit from the base object, they get registered and the rest works as normal. Creating a namespaced API, implies having magic code to deal that (monkey-patching namespaces).

3. On the various ways to modify the `ggplot` object, i.e

   ```python
   plot.geom('point', color='red')
   plot.scale('x', 'continuous', limits=(3,5)))
   ```

   The drawback is, you would still have the classes that do the implementations, and that would make the mapping between the usage and documentation less obvious.

   ```python
   plot.geom(gg.geom.point, color='red')
   plot.geom(gg.geom.point(color='red')

   This is more verbose than `plot.geom_point(color='red')`

   When you have various ways to do stuff, the code that users encounter online has a higher likelihood of looking less familiar. A good example is how users get confused by Matplotlib's 2 plotting APIs.

4. Adding a list is really cool, you can do it already using the `Layers` class.
   e.g

   ```python
   plot + Layers([geom_point(), stat_smooth()])
   ```

   I think it would be easier if you did not have to wrap the list first.

   ```python
   plot + [geom_point(), stat_smooth()]
   ```

   But given the changes that may result from this discussion it would be more like:

   ```python
   import plotnine as p9
   from plotnine import ggplot

   plot.Layers([p9.geom_point(), p9.stat_smooth()])
   ```

   Two imports!

5. I do not think datasets are a big issue.

6. On changing the computed aesthetics; as I understand your suggestion, I do not think it would work and I also do not think it is a good idea. Mapping the same aesthetic to the initial data and to the computed data, *those are really two layers*. Each `geom` or `stat` call creates a single layer, otherwise you violate of the Layered Grammar.

I agree, @chris-b1's chaining suggestion is very convincing and the implementation would look a lot like @azog0's.

**Comment by ChickenProp** [comment]
> On changing the computed aesthetics; as I understand your suggestion, I do not think it would work and I also do not think it is a good idea. Mapping the same aesthetic to the initial data and to the computed data, those are really two layers. Each geom or stat call creates a single layer, otherwise you violate of the Layered Grammar.

Hm, I think I didn't explain this very well. You're only making one layer. As an example, consider `stat_smooth(aes(x='carat', y='price'))`. The smoothing function takes as input a dataframe with x and y columns, and returns one with x, y, ymin, ymax and se. Then geom_smooth uses x, y, ymin and ymax to draw a line plus a ribbon.

Right now you can fiddle with the ribbon with calculated aesthetics, which change the dataframe between the stat and the geom. So if you want to double its width, you can add to the aes something like `ymax='..y + se*2..', ymin='..y - se*2..'`. And then geom_smooth sees different values for ymin and ymax.

But right now you can't fiddle with the line with calculated aesthetics. Suppose you wanted to make the line run along the top edge of the ribbon. You could try to add `y='..ymax..'` to the aes. But that conflicts with the original `y='price'`; the full mapping is only `aes(x='carat', y='..ymax..'). The smoothing function doesn't see a y at all, and can't run.

But I think there's no inherent reason why calculated aesthetics have to conflict with non-calculated ones. In gragrapy, I was giving them separate names. So for the double-width ribbon, you'd do `aes(x='carat', y='price', stat_ymin='y + se*2', stat_ymax='y - se*2')`. And for the line on the top edge, you'd do `aes(x='carat', y='price', stat_y='ymax')`.

**Comment by ChickenProp** [comment]
(Separate comment because these are orthogonal.)

> I do not think datasets are a big issue.

Yeah, this is totally minor.

Regarding the chaining API: I really don't want to bikeshed, so I think you should go ahead and implement whatever you think is best and not feel the need to justify your decision. The rest of this comment is unimportant.

That said,

>     plot.geom('point', color='red')
>     plot.scale('x', 'continuous', limits=(3,5)))

> The drawback is, you would still have the classes that do the implementations, and that would make the mapping between the usage and documentation less obvious.

That's true, but I think it's already the case with `stat='identity'` and `position='dodge'`.

And I do think there's value in being able to add a geom/stat by name. For example, a thing I forgot to mention that gragrapy had was a rudimentary command-line interface. That had to take command-line arguments like `-g point` and turn them into `add geom.point` to the plot, which is really easy if plot `plot.geom('point')` works.

But I don't think this is common enough that the API needs to be designed around it; if there's an interface like `p9.make_geom('point', color='red')`, or `p9.find_geom('point')(color='red')`, that would be fine. (And something similar for stats and scales.) And even if there's no official way to do it, it's not hard to add it as a hack.

(This is similar to what `p9.geoms.geom.geom.from_stat()` does, but that assumes you already have a stat.)

>     plot.geom(gg.geom.point, color='red')
>     plot.geom(gg.geom.point(color='red'))

> This is more verbose than `plot.geom_point(color='red')`

True. The advantage that it gives is if a user implements their own geom, they can do `plot.geom(my_geom)` instead of falling back on something else. But again, this probably isn't common enough that the API needs to be designed around it.

> When you have various ways to do stuff, the code that users encounter online has a higher likelihood of looking less familiar. A good example is how users get confused by Matplotlib's 2 plotting APIs.

Absolutely. I think this is less of an issue with "function that does one thing, but accepts many styles of argument" than with "multiple functions that do the same thing in different ways", but its a consideration.

>     import plotnine as p9
>     from plotnine import ggplot
>
>     plot.Layers([p9.geom_point(), p9.stat_smooth()])

I think what I'd suggest here is simply `plot.add([p9.geom_point(), p9.stat_smooth()])`, where `.add()` simply returns `self + ...`. (And either `.add` or `.__add__` has some code to handle lists, before in turn calling `.__radd__` on the list items.) That seems more general; you could also use it for `plot.add([p9.geom_point(), p9.scale_y_sqrt()])`.

So to summarize: the API I designed lets you do

    plot.geom('point', color='red') # useful if 'point' is in a variable
    plot.geom(my_geom, color='red') # useful for custom geoms

but these are fairly marginal use-cases, and I think that

    plot.add(p9.make_geom('point', color='red'))
    plot.add(my_geom(color='red'))

are almost as good.

And once again, I mostly think you should do whatever you like best, and not bother justifying it if you don't feel like it. I'm happy with any of the options.

**Comment by shoyer** [comment]
@has2k1 are you proposing that third-party libraries implementing a geom or facet would also register methods for chaining on the base `ggplot` object? e.g., `ggplot(...).geom_third_party(...)` could work? This seems pretty reasonable to me, but be aware that you are making a [maintainability tradeoff](https://beam.apache.org/blog/2016/05/27/where-is-my-pcollection-dot-map.html) here.

**Comment by has2k1** [comment]
@shoyer, yes it is the case third-party objects would automatically have first class support. I do have some reservations about having (or just the appearance of) many methods on a class, however I also do not like the only solution I have in mind i.e.

```python
(ggplot(...)
 .add.geom_point(...)
 .add.geom_line(...)
)
```

An intermediary object handicaps the API.

**Comment by has2k1** [comment]
I'm preparing to release v0.3.0, and I cannot resolve this issue. While I am partial to the @jaccovanschaik's  sentiment towards the `*` import, when using `p9` as prefix I have found that the readability of the newer and more complex examples to suffer.

I've also held back on the chainable API, as when properly implemented the members and properties of the main (ggplot) object are delegated to 2nd class citizens, otherwise they get shadowed.

**Comment by shoyer** [comment]
There is no easy solution here, but the Python community does frown pretty heavily upon `import *`. Even for libraries where there are lots of function calls (e.g., NumPy), we usually pick a short prefix. I doubt you could do any better than `import plotnine as p9`. Of course, you are welcome to use whatever you like for your examples, but there is value in establishing a convention for plotnine users.

**Comment by has2k1** [comment]
> Python community does frown pretty heavily upon `import *`

Me too. However with this package & [plydata](https://github.com/has2k1/plydata), I tolerate it in the REPL & Notebook and a solution I'm yet to try would be tailored to that environment. The solution would be to run the code in a separate context/namespace. Something like:

```python
with plotnine_namespace():
    g = ggplot(df, aes('x', 'x')) + geom_point()
    print(g)
```

And/Or using ipython cell magics. 

```python
%%plotnine_namespace

g = ggplot(df, aes('x', 'x')) + geom_point()
print(g)
```

**Comment by shoyer** [comment]
I don't think you can make context-manager namespaces work in a reliable way. In particular, overwriting/restoring locals and non-locals is either impossible or requires CPython specific trickery.

IPython cell magics might be viable as an option: at least that way it's clear what you're doing is not standard Python. Actually, some sort of cell-specific import statement (e.g., `%%cell_import * from plotnine`) might be more broadly useful, it could be worth inquiring with the IPython folks about that.
